lib/backend/migration/migration.go

```diff
@@ -0,0 +1,196 @@
+	src      backend.Backend
+	dst      backend.Backend
+	parallel int
+	migrated atomic.Int64
+	log      logrus.FieldLogger
+}
+
+// MigrationConfig configures a [Migration] with a source and destination backend.
+// All items from the source are copied to the destination. All Teleport Auth
+// Service instances should be stopped when running a migration to avoid data
+// inconsistencies.
+type MigrationConfig struct {
+	// Source is the backend [backend.Config] items are migrated from.
+	Source backend.Config `yaml:"src"`
+	// Destination is the [backend.Config] items are migrated to.
+	Destination backend.Config `yaml:"dst"`
+	// Parallel is the number of items that will be migraated in parallel.
+	Parallel int `yaml:"parallel"`
+	// Log logs the progress of a [Migration]
+	Log logrus.FieldLogger
+}
+
+// New returns a [Migration] based on the provided [MigrationConfig].
+func New(ctx context.Context, config MigrationConfig) (*Migration, error) {
+	src, err := backend.New(ctx, config.Source.Type, config.Source.Params)
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to create source backend")
+	}
+	dst, err := backend.New(ctx, config.Destination.Type, config.Destination.Params)
+	if err != nil {
+		return nil, trace.Wrap(err, "failed to create destination backend")
+	}
+	migration := &Migration{
+		src:      src,
+		dst:      dst,
+		parallel: config.Parallel,
+		log:      config.Log,
+	}
+	if migration.parallel <= 0 {
+		migration.parallel = 1
+	}
+	if migration.log == nil {
+		migration.log = logrus.WithField(teleport.ComponentKey, "migration")
+	}
+	return migration, nil
+}
+
+func (m *Migration) Close() error {
+	var errs []error
+	if m.src != nil {
+		err := m.src.Close()
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+	if m.dst != nil {
+		err := m.dst.Close()
+		if err != nil {
+			errs = append(errs, err)
+		}
+	}
+	return trace.NewAggregate(errs...)
+}
+
+// Run runs a [Migration] until complete.
+func (m *Migration) Run(ctx context.Context) error {
+	itemC := make(chan backend.Item, bufferSize)
+	start := backend.Key("")
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	putGroup, putCtx := errgroup.WithContext(ctx)
+	putGroup.SetLimit(m.parallel)
+
+	getGroup, getCtx := errgroup.WithContext(ctx)
+	getGroup.Go(func() error {
+		for {
+			var result *backend.GetResult
+			defer close(itemC)
+			err := retry(getCtx, 3, func() error {
+				var err error
+				result, err = m.src.GetRange(getCtx, start, backend.RangeEnd(start), bufferSize)
+				if err != nil {
+					return trace.Wrap(err)
+				}
+				return nil
+			})
+			if err != nil {
+				return trace.Wrap(err)
+			}
+			for _, item := range result.Items {
+				select {
+				case itemC <- item:
+				case <-getCtx.Done():
+					return trace.Wrap(getCtx.Err())
+				// This case indicates no consumers are pulling items
+				// from the channel. Return to avoid deadlock.
+				case <-putCtx.Done():
+					return trace.Wrap(putCtx.Err())
+				}
+			}
```

Comment:
Oh right I think that might work. Originally I was worried the puts might consume all of the available go routines but since get is called first that should be fine. I'll just make sure to add 1 to the goroutine limit so its always at least 2.

Reply:
Moved to single errgroup [27e326b](https://github.com/gravitational/teleport/pull/41866/commits/27e326b7b6004788c0dfab989c873a49d420b8f1)